### PR TITLE
Test shutdown

### PR DESCRIPTION
**Pull request by cachedout**
This still isn't 100% complete but it does make some improvements.

This modified our implementation of Popen to kill child procs if they are found. This obviates the need to collect child procs in the caller and kill them.

Secondly, this implements a customized version of the multiprocess poller, which is more understanding about processes which may have been reaped underneath it. This is somewhat of a stopgap measure.

 Finally, this removes the need to send a `.communicate()` to the process, since we aren't processing the output anyhow. 

cc: @s0undt3ch

### CODE REVIEW

File: tests/integration/__init__.py [hunk 1]
```diff
@@ -554,7 +554,7 @@ def terminate(self):
         self._connectable.clear()
         time.sleep(0.0125)
         terminate_process_pid(self._process.pid)
-        self._process.join()
+#        self._process.join()
```
Comment (cachedout) [inline tests/integration/__init__.py]: I don't like removing this but at present, doing so produces a stacktrace and prevents the unit tests from running.

File: salt/utils/nb_popen.py [hunk 2]
```diff
@@ -96,6 +96,25 @@ def __init__(self, *args, **kwargs):
             args if logging_command is None else logging_command
         )
 
+    def poll(self, flag=os.WNOHANG):
+        if self.returncode is None:
+            notintr = False
+            while notintr:
```
Comment (s0undt3ch) [inline salt/utils/nb_popen.py]: This `while` won't run since `notintr` starts as false, right?

Comment (s0undt3ch) [inline salt/utils/nb_popen.py]:
```python
while notintr is False:
   ...
```
?

File: tests/integration/__init__.py (hunk 1 above)
Comment (s0undt3ch) [inline tests/integration/__init__.py]: Produces a stackstrace *with* the `.join()` right?

Comment (cachedout) [inline tests/integration/__init__.py]: Yes.

File: salt/utils/nb_popen.py (hunk 2 above)
Comment (cachedout) [inline salt/utils/nb_popen.py]: I hate it when I forget to remove my testing. :]

Fixed pushed. Thanks.

**Review by s0undt3ch — APPROVED** [review]
Looks Good!

**Comment by s0undt3ch** [comment]
Actually, needs lint fixing ;)

**Comment by cachedout** [comment]
I am a little bit inclined to hold off on this and see if the recent changes in #36555 improve this situation at all.

**Pull request merged**